java/src/main/native/src/RmmJni.cpp

```diff
@@ -66,8 +85,42 @@ class java_event_handler_memory_resource final : public device_memory_resource {
   JavaVM* jvm;
   jobject handler_obj;
   jmethodID on_alloc_fail_method;
+  jmethodID on_alloc_threshold_method;
+  jmethodID on_dealloc_threshold_method;
+
+  // sorted memory thresholds to trigger callbacks
+  std::vector<std::size_t> alloc_thresholds{};
+  std::vector<std::size_t> dealloc_thresholds{};
+
+  std::size_t total_allocated{0};
+
+  // map nd associated lock to track memory sizes by address
+  // TODO: This should be removed when rmm::alloc and rmm::free are removed and the size parameter
+  //       for do_deallocate can be trusted. If map and mutex are removed then total_allocated
+  //       should be updated to be atomic.
+  std::unordered_map<void*, std::size_t> size_map{};
+  std::mutex size_map_mutex{};
 
-  bool on_alloc_fail(std::size_t bytes) {
+
+  static void update_thresholds(JNIEnv* env, std::vector<std::size_t>& thresholds, jlongArray from_java) {
+    thresholds.clear();
+    if (from_java != nullptr) {
+      cudf::jni::native_jlongArray jvalues(env, from_java);
+      thresholds.insert(thresholds.end(), jvalues.data(), jvalues.data() + jvalues.size());
+    } else {
+      // use a single, maximum-threshold value so we don't have to always check for the corner case.
+      thresholds.push_back(std::numeric_limits<std::size_t>::max());
+    }
+  }
+
+  static void log_callback_exception(JNIEnv* env, char const* callback_name, jthrowable throwable) {
+    std::cerr << callback_name << "  threw a Java exception: ";
```

Comment:
Isn't log and rethrow an anti-pattern?

Reply:
Yes, but in this case I am worried about the exception getting clobbered by another exception being thrown from C++, generating a new Java exception, and the details of this exception getting lost.  Any exception being thrown from the allocator will get translated into a `bad_alloc` being thrown by RMM.  That in turn will be caught by the JNI and translated into an OOM, and the original exception details are lost.

Reply:
okay I am +1 for now

Reply:
Thought about this some more.  Would it make sense to *not* throw a new Java exception when a `bad_alloc` is caught if the JVM already has an exception pending for that thread?  Then we could throw in the callback handling code and keep it from being clobbered by the `bad_alloc` processing.  Not sure if this causes a problem in another scenario though.

Reply:
Based on our offline conversation, I went ahead and removed the log-then-throw and instead have the exception handling avoid posting a new exception if one is already pending.  I also added a test to verify the exceptions thrown by the handler are being propagated properly.